### ISSUE TITLE
Handling KubernetesClientException when the DNS server is not available

### ISSUE THREAD

**Issue by ppatierno** [strimzi/strimzi#241]
If the Kubernetes DNS server is not available, the cluster controller (but in general any other component connecting to the API server, so the topic controller as well) raises the following exception which we have to handle in some way (for example exiting and leave Kubernetes restarting the container).

Exception in thread "main" io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:62)
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:53)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getRootPaths(BaseOperation.java:243)
	at io.fabric8.kubernetes.client.BaseClient.rootPaths(BaseClient.java:140)
	at io.fabric8.openshift.client.OpenshiftAdapterSupport.isOpenShift(OpenshiftAdapterSupport.java:53)
	at io.fabric8.openshift.client.OpenshiftAdapterSupport.isAdaptable(OpenshiftAdapterSupport.java:37)
	at io.fabric8.kubernetes.client.BaseClient.isAdaptable(BaseClient.java:122)
	at io.strimzi.controller.cluster.Main.run(Main.java:54)
	at io.strimzi.controller.cluster.Main.main(Main.java:37)
Caused by: java.net.UnknownHostException: kubernetes.default.svc
	at java.net.InetAddress.getAllByName0(InetAddress.java:1280)
	at java.net.InetAddress.getAllByName(InetAddress.java:1192)
	at java.net.InetAddress.getAllByName(InetAddress.java:1126)
	at okhttp3.Dns$1.lookup(Dns.java:39)
	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:185)
	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:149)
	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:84)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:213)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:134)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:113)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.HttpClientUtils$2.intercept(HttpClientUtils.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:377)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:343)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:327)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getRootPaths(BaseOperation.java:236)
	... 6 more

**Comment by tombentley** [comment]
This can be tested on minikube by deleting the kube-dns deployment while performing various operations (such as creating the CC, creating a kafka cluster, creating a kafka topic).

### No `kube-dns` when CC starts

I found no problems when doing `kubectl delete deployment --namespace=kube-system kube-dns && kubectl create -f examples/install/cluster-controller/`.

### No `kube-dns` on Cluster CM Creation

Doing `kubectl delete deployment --namespace=kube-system kube-dns && kubectl create -f examples/configmaps/cluster-controller/kafka-ephemeral.yaml` I found:

* The zk pod took a long time to come up, but eventually it did
* The following NPE in the CC (which should be fixed by #264):

```
2018-03-13 12:14:46 INFO  ClusterController:81 - Triggering periodic reconciliation for namespace default...
2018-03-13 12:14:46 INFO  AbstractClusterOperations:237 - Reconciling Kafka clusters ...
2018-03-13 12:14:46 INFO  AbstractClusterOperations:247 - Reconciliation: Kafka cluster my-cluster should be checked for updates
2018-03-13 12:14:46 INFO  AbstractClusterOperations:248 - Checking for updates in Kafka cluster my-cluster
2018-03-13 12:14:46 INFO  KafkaClusterOperations:615 - Updating Zookeeper cluster my-cluster-zookeeper in namespace default
2018-03-13 12:14:46 INFO  AbstractClusterOperations:148 - update zookeeper cluster my-cluster-zookeeper in namespace default
2018-03-13 12:14:46 INFO  AbstractClusterOperations:158 - zookeeper cluster my-cluster-zookeeper in namespace default: successful update
2018-03-13 12:14:46 INFO  KafkaClusterOperations:436 - Updating Kafka cluster my-cluster-kafka in namespace default
2018-03-13 12:14:46 ERROR AbstractClusterOperations:150 - Error while getting required kafka cluster state for update operation
java.lang.NullPointerException
	at io.strimzi.controller.cluster.resources.KafkaCluster.diff(KafkaCluster.java:210)
	at io.strimzi.controller.cluster.operations.cluster.KafkaClusterOperations$7.getCluster(KafkaClusterOperations.java:439)
	at io.strimzi.controller.cluster.operations.cluster.AbstractClusterOperations.execute(AbstractClusterOperations.java:147)
	at io.strimzi.controller.cluster.operations.cluster.KafkaClusterOperations.lambda$update$5(KafkaClusterOperations.java:692)
	at io.strimzi.controller.cluster.operations.cluster.AbstractClusterOperations.lambda$execute$0(AbstractClusterOperations.java:159)
	at io.vertx.core.impl.FutureImpl.setHandler(FutureImpl.java:81)
	at io.strimzi.controller.cluster.operations.cluster.AbstractClusterOperations.execute(AbstractClusterOperations.java:156)
	at io.strimzi.controller.cluster.operations.cluster.KafkaClusterOperations.update(KafkaClusterOperations.java:688)
	at io.strimzi.controller.cluster.operations.cluster.AbstractClusterOperations.lambda$reconcileCluster$5(AbstractClusterOperations.java:249)
	at io.vertx.core.shareddata.impl.AsynchronousLock.lambda$lockAcquired$0(AsynchronousLock.java:66)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
2018-03-13 12:14:46 ERROR AbstractClusterOperations:253 - Failed to update Kafka cluster my-cluster.
```

* No kafka SS ever gets created.
* If I subsequently try to delete the cluster CM: 
    * I get the following NPE (Again, this should be fixed by #264):

```
2018-03-13 12:22:33 INFO  ClusterController:142 - ConfigMap my-cluster in namespace default was DELETED
2018-03-13 12:22:33 INFO  AbstractClusterOperations:237 - Reconciling Kafka clusters ...
2018-03-13 12:22:33 INFO  AbstractClusterOperations:275 - Reconciliation: Kafka cluster my-cluster-zookeeper should be deleted
2018-03-13 12:22:33 INFO  AbstractClusterOperations:277 - Deleting Kafka cluster my-cluster in namespace default
2018-03-13 12:22:33 ERROR AbstractClusterOperations:150 - Error while getting required kafka cluster state for delete operation
java.lang.NullPointerException
	at io.strimzi.controller.cluster.resources.Labels.fromResource(Labels.java:110)
	at io.strimzi.controller.cluster.resources.KafkaCluster.fromStatefulSet(KafkaCluster.java:157)
	at io.strimzi.controller.cluster.operations.cluster.KafkaClusterOperations$4.getCluster(KafkaClusterOperations.java:292)
	at io.strimzi.controller.cluster.operations.cluster.AbstractClusterOperations.execute(AbstractClusterOperations.java:147)
	at io.strimzi.controller.cluster.operations.cluster.KafkaClusterOperations.deleteKafkaAndZookeeper(KafkaClusterOperations.java:390)
	at io.strimzi.controller.cluster.operations.cluster.KafkaClusterOperations.delete(KafkaClusterOperations.java:384)
	at io.strimzi.controller.cluster.operations.cluster.AbstractClusterOperations.lambda$reconcileCluster$5(AbstractClusterOperations.java:281)
	at io.vertx.core.shareddata.impl.AsynchronousLock.lambda$lockAcquired$0(AsynchronousLock.java:66)
	at io.vertx.core.impl.ContextImpl.lambda$wrapTask$2(ContextImpl.java:344)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:748)
2018-03-13 12:22:33 ERROR AbstractClusterOperations:285 - Failed to delete Kafka cluster my-cluster in namespace default
```

    * The zookeeper cluster is not deleted, but can be deleted manually.

### No `kube-dns` when TC starts

`kubectl delete deployment --namespace=kube-system kube-dns && kubectl delete po my-cluster-topic-controller-6cc5d4fdc6-xdt4s` cause the TC to log a few exceptions, but eventually it came up and did a periodic reconciliation.

### No `kube-dns` when Topic CM created

`kubectl delete deployment --namespace=kube-system kube-dns && kubectl create -f examples/configmaps/topic-controller/kafka-topic-configmap.yaml` didn't seem to cause any problems.

**Comment by scholzj** [comment]
The Cluster controller behaves as expected: When the DNS is not available and it cannot connect to the Kubernetes API, it will loop in restarts (which is what we want). If it looses the DNS while already running, the connection tot he Kubernetes API remains working.
